### PR TITLE
feat!: interchain accounts

### PR DESCRIPTION
**Pull request by rootulp**
Closes https://github.com/celestiaorg/celestia-app/issues/3062
Opens https://github.com/celestiaorg/celestia-app/issues/3207

### CODE REVIEW

**Comment by rootulp** [comment]
Docker image was pushed to https://github.com/celestiaorg/celestia-app/pkgs/container/celestia-app/191669081?tag=pr-3182

**Comment by rootulp** [comment]
```
    ica_test.go:75:
        	Error Trace:	/Users/rootulp/git/rootulp/celestiaorg/celestia-app/interchaintest/ica_test.go:75
        	Error:      	"[]" should have 1 item(s), but has 0
        	Test:       	TestICA
```

**Comment by rootulp** [comment]
```
        -2024-03-15T19:56:10.332900Z	debug	Error querying block data	{"chain_name": "celestia", "chain_id": "celestia", "height": 2, "error": "error in json rpc client, with http response metadata: (Status: 200 OK, Protocol HTTP/1.1). RPC error -32603 - Internal error: node is not persisting abci responses"}
```

**Comment by rootulp** [comment]
```
    ica_test.go:103:
        	Error Trace:	/Users/rootulp/git/rootulp/celestiaorg/celestia-app/interchaintest/ica_test.go:103
        	Error:      	Received unexpected error:
        	            	exit code 1:  Error: unknown command "intertx" for "tx"
```

**Comment by rootulp** [comment]
If we want the ICA tests to run on every PR, we're going to need to build Docker images on every PR and then plumb the appropriate label (PR # or commit hash) into `make test-interchain`. Since we currently only build Docker images for branches on the upstream repo (a.k.a not on forks), we could alternatively only run this test on new commits to `main` which should produce new Docker images already.

**Comment by rootulp** [comment]
I'll pause work on the `interchaintest`s because I'm making extremely slow progress. I'll try to get this PR to a review-able state _without_ the end to end tests and then try adding `interchaintest`s in a follow-up PR.

**Review by evan-forbes — APPROVED** [review]
nice!

**Review by evan-forbes — DISMISSED** [review]

File: app/default_overrides.go [hunk 1]
```diff
@@ -149,6 +151,21 @@ func (ibcModule) DefaultGenesis(cdc codec.JSONCodec) json.RawMessage {
...
+
+// DefaultGenesis returns custom ica module genesis state.
+func (icaModule) DefaultGenesis(cdc codec.JSONCodec) json.RawMessage {
+	gs := icagenesistypes.DefaultGenesis()
+	gs.HostGenesisState.Params.AllowMessages = icaAllowMessages()
+	gs.HostGenesisState.Params.HostEnabled = true
+	gs.ControllerGenesisState.Params.ControllerEnabled = false
+	return cdc.MustMarshalJSON(gs)
+}
+
```
Comment (cmwaters) [inline app/default_overrides.go:168]: [No action needed] It's interesting that we've had to develop this pattern of wrapping the modules. It makes me think that the default genesis should be provided in the constructor

File: app/test/ica_test.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+	})
+	testApp.BeginBlock(abci.RequestBeginBlock{Header: tmproto.Header{
+		Height:  2,
+		Version: version.Consensus{App: 1},
+	}})
+
+	// app version should not have changed yet
+	require.EqualValues(t, 1, testApp.AppVersion())
+
+	// global min gas price should not have been set yet
```
Comment (cmwaters) [inline app/test/ica_test.go:44]: I think this comment needs to be updated

File: app/test/ica_test.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+	newCtx := testApp.NewContext(true, tmproto.Header{
+		Version: version.Consensus{
+			App: 2,
+		},
+	})
+
+	got, err := testApp.ParamsKeeper.Params(newCtx, &proposal.QueryParamsRequest{
+		Subspace: icahosttypes.SubModuleName,
+		Key:      string(icahosttypes.KeyHostEnabled),
+	})
```
Comment (cmwaters) [inline app/test/ica_test.go:70]: nit: Does the ICA have it's own grpc method for getting the params? It's probably better to query that (as it should set the context for you and return all params, not just the `HostEnabled` one

**Review by cmwaters — APPROVED** [review]
Everything looks like it's correctly plumbed here. Just a few minor nits

File: app/default_overrides.go (hunk 1 above)
Comment (rootulp) [reply app/default_overrides.go:168]: TBH I also thought this pattern was a bit strange but I didn't see another way to override default genesis params.

File: app/test/ica_test.go (hunk 2 above)
Comment (rootulp) [reply app/test/ica_test.go]: 6c016b4a6cc730d29e8834e0e293d54f75bc3d8b

File: app/test/ica_test.go (hunk 3 above)
Comment (rootulp) [reply app/test/ica_test.go:69]: ICA does have a GRPC method for params:

Screenshot 2024-04-02 at 1 53 58 PM

I just can't find code that queries params via GRPC from the testApp. Will attempt to switch from ParamsKeeper to GRPC method

File: app/test/ica_test.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
+package app_test
```
[the remark this answers is not in the document]
Comment (rootulp) [reply app/test/ica_test.go]: 08044467de5358b39ea1c51652eeab0d1ea23901

File: app/test/ica_test.go (hunk 3 above)
Comment (rootulp) [reply app/test/ica_test.go]: Done

**Comment by rootulp** [comment]
Oof it looks like the testApp doesn't spin up a GRPC server so the test was only passing locally because I have a celestia-app running that exposes the GRPC server on localhost. Will prob revert back to just querying via the param keeper b/c I can't find an example where testApp spins up a GRPC server.

**Review by evan-forbes — APPROVED** [review]
very nice

**Comment by rootulp** [comment]
Gahhhh another merge conflict 💀

**Review by ninabarbakadze — DISMISSED** [review]
thanks Rootul and sorry for the million merge conflicts

**Review by cmwaters — DISMISSED** [review]
Looks like everything is plumbed together correctly. Also sorry for causing those merge conflicts 🙃

**Review by evan-forbes — APPROVED** [review]

**Review by cmwaters — APPROVED** [review]

**Pull request merged**